libopenage/presenter/presenter.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019-2023 the openage authors. See copying.md for legal info.
+
+#include "presenter.h"
+
+#include <eigen3/Eigen/Dense>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "engine/engine.h"
+#include "log/log.h"
+#include "renderer/camera/camera.h"
+#include "renderer/entity_factory.h"
+#include "renderer/gui/gui.h"
+#include "renderer/gui/integration/public/gui_application_with_logger.h"
+#include "renderer/gui/qml_info.h"
+#include "renderer/resources/shader_source.h"
+#include "renderer/resources/texture_info.h"
+#include "renderer/stages/screen/screen_renderer.h"
+#include "renderer/stages/skybox/skybox_renderer.h"
+#include "renderer/stages/terrain/terrain_renderer.h"
+#include "renderer/stages/world/world_renderer.h"
+#include "renderer/window.h"
+#include "util/path.h"
+
+namespace openage::presenter {
+
+Presenter::Presenter(const util::Path &root_dir,
+                     const std::shared_ptr<engine::Engine> engine) :
+	root_dir{root_dir},
+	render_passes{},
+	engine{engine} {}
+
+
+void Presenter::run() {
+	log::log(INFO << "presenter launching...");
+
+	this->init_graphics();
+
+	if (this->engine) {
+		auto render_factory = std::make_shared<renderer::RenderFactory>(this->terrain_renderer,
+		                                                                this->world_renderer);
+		this->engine->attach_renderer(render_factory);
+	}
+
+	while (not this->window->should_close()) {
+		this->gui_app->process_events();
+		// this->gui->process_events();
+
+		this->render();
+
+		this->renderer->check_error();
+
+		this->window->update();
+	}
+	log::log(MSG(info) << "draw loop exited");
+
+	if (this->engine) {
+		engine->stop();
+	}
+
+	this->window->close();
+}
+
+void Presenter::set_engine(const std::shared_ptr<engine::Engine> &engine) {
```

Comment:
why does the engine need to be set? the other way round seems to be better - first create the engine, then the presenter, then fetch the renderer from the presenter and attach it to the engine.
same goes for the `::run` implementation, it should have side effects of configuring the engine.

Reply:
what you describe is how it is already implemented currently (when starting from `main`).

i added `set_engine` because i though it might be beneficial for testing the presenter if the engine is either optional or can be attached later.